src/modules/Elsa.Workflows.Core/Activities/Flowchart/Activities/Flowchart.cs

```diff
@@ -325,4 +326,55 @@ private async ValueTask OnActivityCanceledAsync(CancelSignal signal, SignalConte
     {
         await CompleteIfNoPendingWorkAsync(context.ReceiverActivityExecutionContext);
     }
+
+    private async ValueTask OnActivityExceptionAsync(ExceptionSignal signal, SignalContext context)
+    {
+        var avtivities = Activities.Where(a => a is Catch).Select(a => a as Catch).ToList();
+
+        if (avtivities.Count == 0)
+            return;
+
+        var flowchartContext = context.ReceiverActivityExecutionContext;
+        var workflowExecutionContext = flowchartContext.WorkflowExecutionContext;
+
+        var exceptionActivity = signal.Activity;
+
+        var options = new ActivityInvocationOptions
+        {
+            Owner = flowchartContext
+        };
+
+        foreach (var catchAvtivity in avtivities)
+        {
+            var catchContext = workflowExecutionContext.ActivityExecutionContexts.FirstOrDefault(x => x.Activity.Id == catchAvtivity!.Id);
+
+            if (catchContext == null)
+            {
+                catchContext = workflowExecutionContext.CreateActivityExecutionContext(catchAvtivity!, options);
+                workflowExecutionContext.AddActivityExecutionContext(catchContext);
+
+                await catchContext.EvaluateInputPropertiesAsync();
+            }
+
+            var scheduleWorkOptions = new ScheduleWorkOptions
+            {
+                CompletionCallback = OnChildCompletedAsync,
+                ExistingActivityExecutionContext = catchContext
+            };
+
+            var catchAll = catchAvtivity!.CatchAll.Get(catchContext);
+            if (catchAll)
+            {
+                var connections = Connections.Descendants(catchAvtivity).ToList();
+                if (!connections.Any(x => x.Target.Activity == exceptionActivity || x.Source.Activity == exceptionActivity))
+                    await flowchartContext.ScheduleActivityAsync(catchAvtivity, scheduleWorkOptions);
+            }
+            else
+            {
+                var catchActivities = catchAvtivity.CatchActivities?.Get(catchContext);
+                if (catchActivities != null && catchActivities.Contains(exceptionActivity.Id))
+                    await flowchartContext.ScheduleActivityAsync(catchAvtivity, scheduleWorkOptions);
+            }
+        }
+    }
```

Comment:
I'm not entirely sure about this approach. Although I understand that you have to evaluate the properties of the `Catch` activity in order to see if the faulting activity ID is present in its list and whether the CatchAll is set to true, I would think that, after it is determined that the CatchAll activity should execute, you should simply schedule the catch activity for execution. I.e, I think that if either `catchAll` is true _or_ the faulting activity is present in the "catch activities" list, you should schedule the catch activity. 

E.g.

```suggestion
        {
            var catchContext = workflowExecutionContext.ActivityExecutionContexts.FirstOrDefault(x => x.Activity.Id == catchAvtivity!.Id);

            if (catchContext == null)
            {
                catchContext = workflowExecutionContext.CreateActivityExecutionContext(catchAvtivity!, options);
                workflowExecutionContext.AddActivityExecutionContext(catchContext);

                await catchContext.EvaluateInputPropertiesAsync();
            }

            var scheduleWorkOptions = new ScheduleWorkOptions
            {
                CompletionCallback = OnChildCompletedAsync,
                ExistingActivityExecutionContext = catchContext
            };

            var catchAll = catchActivity!.CatchAll.Get(catchContext);
            var catchActivities = catchActivity.CatchActivities?.Get(catchContext);
            var catchFaultingActivity = catchActivities?.Contains(exceptionActivity.Id) == true;
            var shouldCatch = catchAll || catchFaultingActivity;
            
             if(shouldCatch) 
                    await flowchartContext.ScheduleActivityAsync(catchActivity, scheduleWorkOptions);
        }
    }
```

Please note that I haven't tested it, so I could be totally off here.

Reply:
When catchAll is true, we need to consider whether abnormal activities are executed after catch activities. Otherwise, an infinite loop may occur.